python/ray/data/_internal/datasource/iceberg_datasink.py

```diff
@@ -3,168 +3,264 @@
         from pyiceberg import catalog
 
         return catalog.load_catalog(self._catalog_name, **self._catalog_kwargs)
 
-    def on_write_start(self) -> None:
-        """Prepare for the transaction"""
-        import pyiceberg
-        from pyiceberg.table import TableProperties
+    def _reload_table(self) -> None:
+        """Reload the Iceberg table from the catalog."""
+        catalog = self._get_catalog()
+        self._table = catalog.load_table(self.table_identifier)
 
-        if version.parse(pyiceberg.__version__) >= version.parse("0.9.0"):
-            from pyiceberg.utils.properties import property_as_bool
-        else:
-            from pyiceberg.table import PropertyUtil
+    def _update_schema(self, incoming_schema: "pa.Schema") -> None:
+        """
+        Update the table schema to accommodate incoming data using union-by-name semantics.
 
-            property_as_bool = PropertyUtil.property_as_bool
+        This is called from the driver after reconciling all schemas.
 
-        catalog = self._get_catalog()
-        table = catalog.load_table(self.table_identifier)
-        self._txn = table.transaction()
-        self._io = self._txn._table.io
-        self._table_metadata = self._txn.table_metadata
-        self._uuid = uuid.uuid4()
-
-        if unsupported_partitions := [
-            field
-            for field in self._table_metadata.spec().fields
-            if not field.transform.supports_pyarrow_transform
-        ]:
-            raise ValueError(
-                f"Not all partition types are supported for writes. Following partitions cannot be written using pyarrow: {unsupported_partitions}."
-            )
-
-        self._manifest_merge_enabled = property_as_bool(
-            self._table_metadata.properties,
-            TableProperties.MANIFEST_MERGE_ENABLED,
-            TableProperties.MANIFEST_MERGE_ENABLED_DEFAULT,
-        )
+        Args:
+            incoming_schema: The PyArrow schema to merge with the table schema
+        """
+        from pyiceberg.exceptions import CommitFailedException
+
+        max_retries = 3
+
+        for attempt in range(max_retries):
+            try:
+                with self._table.update_schema() as update:
+                    update.union_by_name(incoming_schema)
+                # Succeeded, reload to get latest table version and exit.
+                self._reload_table()
+                return
+            except CommitFailedException:
+                if attempt < max_retries - 1:
+                    logger.debug(
+                        f"Schema update conflict - reloading and retrying "
+                        f"(attempt {attempt + 1}/{max_retries})"
+                    )
+                    self._reload_table()
+                else:
+                    logger.error(
+                        "Failed to update schema after %d retries due to conflicts.",
+                        max_retries,
+                    )
+                    raise
```

Comment:
If we want to be retrying txns we need to be consistent, not sure it makes sense to do just for the schema updates